### PR TITLE
Add extendWhere, orWhere and andWhere methods to database query class.

### PR DESCRIPTION
**Pull request by photodude**
Brings work from the CMS by @chrisdavenport over to the Framework.

CMS PR's https://github.com/joomla/joomla-cms/pull/8718 and https://github.com/joomla/joomla-cms/pull/5837

Testing instructions and details listed in https://github.com/joomla/joomla-cms/pull/5837

### CODE REVIEW

File: src/DatabaseQuery.php [hunk 1]
```diff
@@ -1396,6 +1396,72 @@ public function where($conditions, $glue = 'AND')
...
+	 *
+	 * Usage:
+	 * $query->where(array('a = 1', 'b = 2'))->extendWhere('XOR', array('c = 3', 'd = 4'));
+	 * will produce: WHERE ((a = 1 AND b = 2) XOR (c = 3 AND d = 4)
+	 *
+	 * @param   string  $outerGlue   The glue by which to join the conditions to the current WHERE conditions.
+	 * @param   mixed   $conditions  A string or array of WHERE conditions.
+	 * @param   string  $innerGlue   The glue by which to join the conditions. Defaults to AND.
+	 *
+	 * @return  JDatabaseQuery  Returns this object to allow chaining.
```
Comment (mbabker) [inline src/DatabaseQuery.php]: `@return  DatabaseQuery  Returns this object to allow chaining.`

File: src/DatabaseQuery.php [hunk 2]
```diff
@@ -1396,6 +1396,72 @@ public function where($conditions, $glue = 'AND')
...
+	 *
+	 * @since   1.3
```
Comment (mbabker) [inline src/DatabaseQuery.php]: `@since   __DEPLOY_VERSION__`

CI will replace the version string when tagged

File: src/DatabaseQuery.php [hunk 3]
```diff
@@ -1396,6 +1396,72 @@ public function where($conditions, $glue = 'AND')
...
+	 */
+	public function extendWhere($outerGlue, $conditions, $innerGlue = 'AND')
+	{
+		// Replace the current WHERE with a new one which has the old one as an unnamed child.
+		$this->where = new JDatabaseQueryElement('WHERE', $this->where->setName('()'), " $outerGlue ");
```
Comment (mbabker) [inline src/DatabaseQuery.php]: `new Query\QueryElement()`

File: src/DatabaseQuery.php [hunk 4]
```diff
@@ -1396,6 +1396,72 @@ public function where($conditions, $glue = 'AND')
...
+
+		// Append the new conditions as a new unnamed child.
+		$this->where->append(new JDatabaseQueryElement('()', $conditions, " $innerGlue "));
```
Comment (mbabker) [inline src/DatabaseQuery.php]: `new Query\QueryElement()`

File: src/Query/QueryElement.php [hunk 5]
```diff
@@ -108,6 +108,22 @@ public function getElements()
 	}
 
 	/**
+	 * Sets the name of this element.
+	 *
+	 * @param   string  $name  Name of the element.
+	 *
+	 * @return  JDatabaseQueryElement  Returns this object to allow chaining.
```
Comment (mbabker) [inline src/Query/QueryElement.php]: Same comments as the other file for the doc blocks

File: Tests/QueryTest.php [hunk 6]
```diff
@@ -1546,6 +1546,15 @@ public function testWhere()
 			'Tests rendered value after second use and array input.'
 		);
 
+		// Add more columns but specify different glue.
+		// Note that the change of glue is ignored.
+		$this->_instance->where(array('faz = 4', 'gaz = 5'), 'OR');
+		$this->assertThat(
+			trim(TestReflection::getValue($this->_instance, 'where')),
```
Comment (mbabker) [inline Tests/QueryTest.php]: Calls to `TestReflection` should be replaced with `TestHelper` (already imported in this file).

File: Tests/QueryTest.php [hunk 7]
```diff
@@ -1564,6 +1573,156 @@ public function testWhere()
...
+	 * Tests the JDatabaseQuery::extendWhere method.
+	 *
+	 * @return  void
+	 *
+	 * @since   1.3
+	 */
+	public function testExtendWhere()
+	{
+		$this->assertThat(
+			$this->_instance->where('foo = 1')->extendWhere('ABC', 'bar = 2'),
```
Comment (mbabker) [inline Tests/QueryTest.php]: `$this->_instance` should be replaced with `$this->instance` (no underscore)

**Comment by photodude** [comment]
Thanks @mbabker I have made the requested corrections and adjustments.

**Pull request closed without merging**